### PR TITLE
Unify the log for vineyardctl and find an appropriate port for forwarding

### PR DESCRIPTION
**Pull request by dashanji**
What do these changes do?
-------------------------

- Unify the log for vineyardctl
- Add a flag `--debug` in the command `vineyardctl ls` for printing debug info.
- Find an appropriate port for forwarding if it's bound.

Related issue number
--------------------

Fixes part of #1503.

### CODE REVIEW

**Comment by sighingnow** [comment]
Use `--log-level={verbose/debug/info/warn/error}` instead.

**Comment by dashanji** [comment]
> Use `--log-level={verbose/debug/info/warn/error}` instead.

Updated.

**Review by sighingnow — APPROVED** [review]

File: k8s/cmd/commands/flags/client_flags.go [hunk 1]
```diff
@@ -105,4 +108,10 @@ func ApplyConnectOpts(cmd *cobra.Command) {
 	cmd.Flags().StringVarP(&DeploymentName, "deployment-name", "", "", "the name of vineyard deployment")
 	cmd.Flags().IntVarP(&Port, "port", "", 9600, "the port of vineyard deployment")
 	cmd.Flags().IntVarP(&ForwardPort, "forward-port", "", 9600, "the forward port of vineyard deployment")
+	cmd.Flags().IntVarP(&Verbose, "log-verbose", "", 0,
+		"the log level to print, default is 0, support:"+"\n"+
+			"- info(0): print info log."+"\n"+
+			"- debug(1): print debug log."+"\n"+
+			"The output will be info log by default, if you set the verbose to 1, "+
+			"the output will be debug log and info log.")
```
Comment (sighingnow) [inline k8s/cmd/commands/flags/client_flags.go:116]: `log-level`, not `log-verbose`. I believe you never see any such naming in other libraries. You can have a simple mapping like

- error -> 5
- warn -> 4
- info -> 3
- debug -> 2
- ....

And just let user to decide "info" / "debug" / "warn" / "error", don't expose the internal number to user.

Comment (dashanji) [reply k8s/cmd/commands/flags/client_flags.go]: Make sense to me.

File: k8s/cmd/README.md [hunk 2]
```diff
@@ -28,6 +28,11 @@ drivers.
   -g, --gen-doc             Generate reference docs, e.g., "./cmd/README.md"
   -h, --help                help for vineyardctl
       --kubeconfig string   kubeconfig path for the kubernetes cluster (default "$HOME/.kube/config")
+      --log-level string    the log level to print, default is info, support:
+                            - info: print info log.
+                            - debug: print debug log.
```
Comment (sighingnow) [inline k8s/cmd/README.md:33]: I think it can also be "warn", "error" and "fatal"?

Comment (sighingnow) [reply k8s/cmd/README.md:33]: etcd --help:

```
 --log-level 'info'
    Configures log level. Only supports debug, info, warn, error, panic, or fatal.
```

argo --help:

```
     --loglevel string                Set the logging level. One of: debug|info|warn|error (default "info")
```

File: k8s/cmd/commands/delete/delete_vineyardd.go [hunk 3]
```diff
@@ -39,9 +39,10 @@ var deleteVineyarddCmd = &cobra.Command{
 	Short:   "Delete the vineyardd cluster from kubernetes",
 	Example: deleteVineyarddExample,
 	Run: func(cmd *cobra.Command, args []string) {
-		util.AssertNoArgs(cmd, args)
+		util.Check(cmd, args)
 		client := util.KubernetesClient()
 
+		log.Debug("deleting Vineyardd cr")
```
Comment (sighingnow) [inline k8s/cmd/commands/delete/delete_vineyardd.go:45]: "deleting Vineyardd cr" -> "deleting Vineyardd cluster"

As you state "delete the vineyard cluster ..." in help message.

File: k8s/cmd/commands/flags/global_flags.go [hunk 4]
```diff
@@ -72,6 +75,12 @@ func ApplyGlobalFlags(cmd *cobra.Command) {
 	cmd.PersistentFlags().
 		BoolVarP(&CreateNamespace, "create-namespace", "", false,
 			"create the namespace if it does not exist, default false")
+	cmd.PersistentFlags().StringVarP(&LogLevel, "log-level", "", "info",
+		"the log level to print, default is info, support:"+"\n"+
+			"- info: print info log."+"\n"+
+			"- debug: print debug log."+"\n"+
+			"If you set the log level to debug, the output will be info log and debug log."+"\n"+
+			"If you set the log level to info, the output will be only info log.")
 	cmd.Flags().BoolVarP(&DumpUsage, "dump-usage", "j", false, "Dump usage in JSON")
```
Comment (sighingnow) [inline k8s/cmd/commands/flags/global_flags.go:84]: Refer to above comment for better help message for `--log-level`.

File: k8s/cmd/commands/util/check.go [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+limitations under the License.
+*/
+package util
+
+import "github.com/spf13/cobra"
+
+func Check(cmd *cobra.Command, args []string) {
+	AssertNoArgs(cmd, args)
+
+	SetLogLevel()
```
Comment (sighingnow) [inline k8s/cmd/commands/util/check.go:23]: Why do you need `SetLogLevel` for every command? Why it cannot be done in main?

File: k8s/cmd/commands/util/log.go [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+func getVerboseFromLevel(level string) int {
+	switch level {
+	case "info":
+		return log.InfoLevel
+	case "debug":
+		return log.Debuglevel
+	default:
+		log.Fatal(nil, "unknown log level")
+		return -1
+	}
```
Comment (sighingnow) [inline k8s/cmd/commands/util/log.go:33]: See the previous comment for verbose / debug / info / warn / error / fatal.

File: go/vineyard/pkg/common/log/log.go [hunk 7]
```diff
@@ -34,26 +34,74 @@ limitations under the License.
...
 	dlog = NewDelegatingLogSink(defaultLogger.GetSink())
 
-	Log = Logger{logr.New(dlog)}
+	Log = Logger{logr.New(dlog).WithName("vineyard")}
 )
 
+func SetVerbose(level int) {
+	defaultLogger = makeDefaultLogger(-level)
+	dlog.Fulfill(defaultLogger.GetSink())
+}
```
Comment (sighingnow) [inline go/vineyard/pkg/common/log/log.go:66]: SetVerbose -> SetLogLevel.

File: k8s/pkg/log/log.go [hunk 8]
```diff
@@ -31,17 +31,27 @@ import (
...
-
-	Log = Logger{logr.New(dlog).WithName("vineyard")}
+	Log           = Logger{logr.New(dlog).WithName("vineyard")}
 )
 
-func makeDefaultLogger() logr.Logger {
+func SetVerbose(level int) {
+	defaultLogger = makeDefaultLogger(-level)
+	dlog.Fulfill(defaultLogger.GetSink())
+}
```
Comment (sighingnow) [inline k8s/pkg/log/log.go:48]: SetVerbose -> SetLogLevel.

**Review by sighingnow — APPROVED** [review]
- log level: debug/info/error/fatal
   - set info/error/fatal: disable v
   - debug: enable v

**Review by sighingnow — APPROVED** [review]
- log.go
   - only keep info/error/fatal
   - alias debug to info
- command line:
   - --verbose: true -> enable info.+ error, false -> only error
- output result: fmt.println

**Pull request merged**